--- dosbox-0.63/include/programs.h~	2004-08-04 11:12:51.000000000 +0200
+++ dosbox-0.63/include/programs.h	2005-12-12 17:37:59.000000000 +0100
@@ -40,7 +40,7 @@
 	CommandLine * cmd;
 	DOS_PSP * psp;
 	virtual void Run(void)=0;
-	bool Program::GetEnvStr(const char * entry,std::string & result);
+	bool GetEnvStr(const char * entry,std::string & result);
 	bool GetEnvNum(Bitu num,std::string & result);
 	Bitu GetEnvCount(void);
 	bool SetEnv(const char * entry,const char * new_string);
