--- dosbox-0.74/include/dos_inc.h~	2010-05-10 19:43:54.000000000 +0200
+++ dosbox-0.74/include/dos_inc.h	2012-02-12 19:37:06.688604486 +0100
@@ -28,6 +28,8 @@
 #include "mem.h"
 #endif
 
+#include <cstddef>
+
 #ifdef _MSC_VER
 #pragma pack (1)
 #endif
--- dosbox-0.74/src/gui/sdlmain.cpp~	2010-05-10 19:43:54.000000000 +0200
+++ dosbox-0.74/src/gui/sdlmain.cpp	2013-11-07 22:23:59.891932656 +0100
@@ -1529,7 +1529,7 @@
 	if ( !sdl.inited && SDL_Init(SDL_INIT_VIDEO|SDL_INIT_NOPARACHUTE) < 0 ) textonly = true;
 	sdl.inited = true;
 #endif
-	printf(message);
+	printf("%s", message);
 	if(textonly) return;
 	if(!sdl.surface) sdl.surface = SDL_SetVideoMode(640,400,0,0);
 	if(!sdl.surface) return;
